pulp_ansible/app/tasks/collections.py

```diff
@@ -100,29 +114,30 @@ async def declarative_content_from_git_repo(remote, url, git_ref=None, metadata_
         gitrepo = Repo.clone_from(url, uuid4(), depth=1, multi_options=["--recurse-submodules"])
     commit_sha = gitrepo.head.commit.hexsha
     metadata, artifact_path = sync_collection(gitrepo.working_dir, ".")
+    artifact = Artifact.init_and_validate(artifact_path)
     if metadata_only:
         metadata["artifact"] = None
         metadata["artifact_url"] = None
     else:
-        artifact = Artifact.init_and_validate(artifact_path)
-        try:
-            await sync_to_async(artifact.save)()
-        except IntegrityError:
-            artifact = Artifact.objects.get(sha256=artifact.sha256)
+        if existing_artifact := await Artifact.objects.filter(sha256=artifact.sha256).afirst():
+            await existing_artifact.atouch()
+            artifact = existing_artifact
+        else:
+            try:
+                assert artifact._state.adding is True
+                await sync_to_async(artifact.save)()
+                assert artifact._state.adding is False
+            except IntegrityError:
+                artifact = await Artifact.objects.aget(sha256=artifact.sha256)
         metadata["artifact_url"] = reverse("artifacts-detail", args=[artifact.pk])
         metadata["artifact"] = artifact
     metadata["remote_artifact_url"] = "{}/commit/{}".format(url.rstrip("/"), commit_sha)
+    metadata["sha256"] = artifact.sha256
 
     artifact = metadata["artifact"]
     try:
-        collection_version = await sync_to_async(create_collection_from_importer)(
-            metadata, metadata_only=metadata_only
-        )
-        await sync_to_async(ContentArtifact.objects.get_or_create)(
-            artifact=artifact,
-            content=collection_version,
-            relative_path=collection_version.relative_path,
-        )
+        collection_version, tags = await sync_to_async(create_collection_from_importer)(metadata)
+        await _save_collection_version(collection_version, artifact, tags)
     except ValidationError as e:
         if "unique" in str(e):
```

Comment:
You need to update the CollectionVersion.get below on line 148 to include the sha256, also can make it use aget.

Reply:
Actually, I think this is the place where we still have the old behaviour, and we should not change that.
Remember this is not fixing the issue, but preparing the database to fix the issue with proper lookup later.